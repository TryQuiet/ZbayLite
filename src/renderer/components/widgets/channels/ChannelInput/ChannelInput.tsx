import React, { useCallback } from 'react'
import classNames from 'classnames'
import ContentEditable, { ContentEditableEvent } from 'react-contenteditable'
import Picker from 'emoji-picker-react'
import Grid from '@material-ui/core/Grid'
import { makeStyles } from '@material-ui/core/styles'
import orange from '@material-ui/core/colors/orange'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import ChannelInputInfoMessage from './ChannelInputInfoMessage'
import { INPUT_STATE } from './InputState.enum'
import Icon from '../../../ui/Icon/Icon'
import emojiGray from '../../../../static/images/emojiGray.svg'
import emojiBlack from '../../../../static/images/emojiBlack.svg'

const useStyles = makeStyles(theme => ({
  root: {
    background: '#fff',
    height: '100%',
    width: '100%'
  },
  '@keyframes blinker': {
    from: { opacity: 0 },
    to: { opacity: 1 }
  },
  input: {
    fontSize: 14,
    outline: 'none',
    padding: '12px 16px',
    lineHeight: '24px',
    '&:empty': {
      '&:before': {
        content: 'attr(placeholder)',
        display: 'block',
        color: '#aaa'
      }
    },
    wordBreak: 'break-word'
  },
  textfield: {
    border: `1px solid ${theme.palette.colors.veryLightGray}`,
    maxHeight: 300,
    'overflow-y': 'auto',
    borderRadius: 4,
    '&:hover': {
      borderColor: theme.palette.colors.trueBlack
    }
  },
  inputsDiv: {
    paddingLeft: '20px',
    paddingRight: '20px',
    width: '100%',
    margin: '0px'
  },
  disabledBottomMargin: {
    marginBottom: 0
  },
  warningIcon: {
    color: orange[500]
  },
  blinkAnimation: {
    animationName: '$blinker',
    animationDuration: '1s',
    animationTimingFunction: 'linear',
    animationIterationCount: 1
  },
  backdrop: {
    height: 'auto',
    padding: `${theme.spacing(1)}px`,
    backgroundColor: 'rgba(0, 0, 0, 0.75)',
    WebkitTapHighlightColor: 'transparent',
    pointerEvents: 'none',
    touchAction: 'none'
  },
  focused: {
    borderColor: theme.palette.colors.trueBlack
  },
  iconButton: {
    marginRight: 0
  },
  highlight: {
    color: theme.palette.colors.lushSky,
    backgroundColor: theme.palette.colors.lushSky12,
    padding: 5,
    borderRadius: 4
  },
  emoji: {
    marginRight: 17,
    marginLeft: 10,
    cursor: 'pointer'
  },
  actions: {
    postion: 'relative'
  },
  picker: {
    position: 'absolute',
    bottom: 60,
    right: 15
  },
  errorIcon: {
    display: 'flex',
    justify: 'center',
    alignItems: 'center',
    marginLeft: 20,
    marginRight: 5
  },
  errorText: {
    color: theme.palette.colors.trueBlack
  },
  errorBox: {
    marginTop: 5
  },
  linkBlue: {
    fontWeight: 'normal',
    fontStyle: 'normal',
    cursor: 'pointer',
    color: theme.palette.colors.linkBlue
  },
  notAllowed: {
    cursor: 'not-allowed'
  }
}))

export interface ChannelInputProps {
  channelAddress: string
  channelName?: string
  channelParticipants?: Array<{ nickname: string }>
  inputPlaceholder: string
  inputState?: INPUT_STATE
  initialMessage?: string
  onChange: (arg: string) => void
  onKeyPress: (input: string) => void
  infoClass: string
  setInfoClass: (arg: string) => void
}

export const ChannelInputComponent: React.FC<ChannelInputProps> = ({
  channelAddress,
  channelParticipants = [],
  inputPlaceholder,
  inputState = INPUT_STATE.AVAILABLE,
  initialMessage = '',
  onChange,
  onKeyPress,
  infoClass,
  setInfoClass
}) => {
  const classes = useStyles({})

  const [_anchorEl, setAnchorEl] = React.useState<HTMLDivElement>(null)
  const [mentionsToSelect, setMentionsToSelect] = React.useState([])

  const messageRef = React.useRef<string>()
  const refSelected = React.useRef<number>()
  const isFirstRenderRef = React.useRef(true)

  const mentionsToSelectRef = React.useRef<any[]>()

  const inputRef = React.createRef<ContentEditable & HTMLDivElement & any>() // any for updater.enqueueForceUpdate

  const [focused, setFocused] = React.useState(false)
  const [selected, setSelected] = React.useState(0)

  const [emojiHovered, setEmojiHovered] = React.useState(false)
  const [openEmoji, setOpenEmoji] = React.useState(false)

  const [htmlMessage, setHtmlMessage] = React.useState<string>(initialMessage)
  const [message, setMessage] = React.useState(initialMessage)

  window.onfocus = () => {
    inputRef?.current?.el.current.focus()
    setFocused(true)
  }

  const scrollToBottom = () => {
    const scroll = document.getElementById('messages-scroll')?.parentElement
    setTimeout(() => {
      if (scroll?.scrollTop) {
        scroll.scrollTop = scroll.scrollHeight
      }
    }, 100)
  }

  React.useEffect(() => {
    inputRef.current.updater.enqueueForceUpdate(inputRef.current)
  }, [inputPlaceholder, channelAddress])

  // Use reference to bypass memorization
  React.useEffect(() => {
    refSelected.current = selected
  }, [selected])

  const isRefSelected = (refSelected: number | undefined): refSelected is number => {
    return typeof refSelected === 'number'
  }

  React.useEffect(() => {
    mentionsToSelectRef.current = mentionsToSelect
  }, [mentionsToSelect])

  React.useEffect(() => {
    if (!message) {
      setHtmlMessage('')
    }
  }, [message])

  React.useEffect(() => {
    setMessage(initialMessage)
    setHtmlMessage(initialMessage)
    if (!isFirstRenderRef.current) {
      return () => {
        if (messageRef?.current) {
          onChange(messageRef.current)
        }
      }
    }
    isFirstRenderRef.current = false
  }, [channelAddress])

  React.useEffect(() => {
    messageRef.current = message
  }, [message])

  const findMentions = React.useCallback(
    (text: string) => {
      // Search for any mention in message string
      const result: string = text.replace(
        /(<span([^>]*)>)?@([a-z0-9]?\w*)(<\/span>)?/gi,
        (match, span, _class, nickname) => {
          // Ignore already established mentions
          if (span?.includes('class')) {
            return match
          }

          nickname = nickname ?? ''
          const possibleMentions = channelParticipants.filter(
            user =>
              user.nickname.startsWith(nickname) &&
              !channelParticipants.find(user => user.nickname === nickname)
          )

          if (JSON.stringify(mentionsToSelect) !== JSON.stringify(possibleMentions)) {
            setMentionsToSelect(possibleMentions)
            setTimeout(() => {
              setSelected(0)
            }, 0)
          }

          // Wrap mention in spans to be able to treat it as an anchor for popper
          return `<span>@${nickname}</span>`
        }
      )

      return result
    },
    [mentionsToSelect, setMentionsToSelect]
  )

  const sanitizedHtml = findMentions(htmlMessage)

  const onChangeCb = useCallback(
    (e: ContentEditableEvent) => {
      if (inputState === INPUT_STATE.AVAILABLE) {
        // @ts-expect-error
        setMessage(e.nativeEvent.target.innerText)
        // @ts-expect-error
        if (!e.nativeEvent.target.innerText) {
          setHtmlMessage('')
        } else {
          setHtmlMessage(e.target.value)
        }
      }
      setAnchorEl(e.currentTarget.lastElementChild)
    },
    [setAnchorEl, onChange, setHtmlMessage]
  )

  const inputStateRef = React.useRef(inputState)
  React.useEffect(() => {
    inputStateRef.current = inputState
  })

  const mentionSelectAction = (e: React.MouseEvent | React.KeyboardEvent) => {
    e.preventDefault()
    const nickname = mentionsToSelectRef.current[refSelected.current].nickname
    setHtmlMessage(htmlMessage => {
      const wrapped = `<span class="${classes.highlight}">@${nickname}</span>&nbsp;`
      return htmlMessage.replace(/<span>[^/]*<\/span>$/g, wrapped)
    })
    // Replace mentions characters with full nickname in original message string
    setMessage(message.replace(/(\b(\w+)$)/, `${nickname} `))
    // Clear popper items after choosing mention
    setMentionsToSelect([])
    inputRef.current?.el.current.focus()
  }

  const onKeyDownCb = useCallback(
    e => {
      if (!isRefSelected(refSelected.current)) {
        throw new Error('refSelected is on unexpected type')
      }
      if (!mentionsToSelectRef?.current) {
        return
      }
      if (mentionsToSelectRef.current.length) {
        if (e.nativeEvent.keyCode === 40) {
          if (refSelected.current + 1 >= mentionsToSelectRef.current.length) {
            setSelected(0)
          } else {
            setSelected(refSelected.current + 1)
          }
          e.preventDefault()
        }
        if (e.nativeEvent.keyCode === 38) {
          if (refSelected.current - 1 < 0) {
            setSelected(mentionsToSelectRef.current.length - 1)
          } else {
            setSelected(refSelected.current - 1)
          }
          e.preventDefault()
        }
        if (e.nativeEvent.keyCode === 13 || e.nativeEvent.keyCode === 9) {
          mentionSelectAction(e)
        }
      }

      if (
        inputStateRef.current === INPUT_STATE.AVAILABLE &&
        e.nativeEvent.keyCode === 13 &&
        e.target.innerText !== ''
      ) {
        e.preventDefault()
        onChange(e.target.innerText)
        onKeyPress(e.target.innerText)
        setMessage('')
        setHtmlMessage('')
        scrollToBottom()
      } else {
        if (e.nativeEvent.keyCode === 13) {
          e.preventDefault()
          if (infoClass !== classNames(classes.backdrop, classes.blinkAnimation)) {
            setInfoClass(classNames(classes.backdrop, classes.blinkAnimation))
            setTimeout(() => setInfoClass(classNames(classes.backdrop)), 1000)
          }
        }
      }
    },
    [
      inputState,
      message,
      mentionsToSelectRef,
      onChange,
      onKeyPress,
      setMessage,
      setHtmlMessage,
      scrollToBottom,
      infoClass,
      setInfoClass,
      setSelected
    ]
  )

  return (
    <Grid
      className={classNames({
        [classes.root]: true,
        [classes.notAllowed]: inputState !== INPUT_STATE.AVAILABLE
      })}>
      <Grid
        container
        className={classNames({
          [classes.root]: true
        })}
        direction='column'
        justify='center'>
        {/* <MentionPoper anchorEl={anchorEl} selected={selected}>
          {mentionsToSelect.map((target, index) => (
            <MentionElement
              key={index}
              name={target.nickname}
              highlight={index === selected}
              onMouseEnter={() => {
                setSelected(index)
              }}
              channelName={channelName}
              onClick={e => {
                mentionSelectAction(e)
              }}
            />
          ))}
        </MentionPoper> */}
        <Grid
          container
          direction='row'
          alignItems='center'
          justify='center'
          spacing={0}
          className={classNames({
            [classes.inputsDiv]: true
          })}>
          <ClickAwayListener
            onClickAway={() => {
              setFocused(false)
            }}>
            <Grid
              item
              xs
              container
              className={classNames({
                [classes.textfield]: true,
                [classes.focused]: focused
              })}
              justify='center'
              alignItems='center'>
              <Grid item xs>
                <ContentEditable
                  ref={inputRef}
                  placeholder={`Message ${inputPlaceholder}`}
                  className={classes.input}
                  onClick={() => {
                    if (!focused) {
                      setFocused(true)
                    }
                  }}
                  html={sanitizedHtml}
                  onChange={onChangeCb}
                  onKeyDown={onKeyDownCb}
                />
              </Grid>
              <Grid item className={classes.actions}>
                <Grid container justify='center' alignItems='center'>
                  <Icon
                    className={classes.emoji}
                    src={emojiHovered ? emojiBlack : emojiGray}
                    onClickHandler={() => {
                      setOpenEmoji(true)
                    }}
                    onMouseEnterHandler={() => {
                      setEmojiHovered(true)
                    }}
                    onMouseLeaveHandler={() => {
                      setEmojiHovered(false)
                    }}
                  />
                </Grid>
                {openEmoji && (
                  <ClickAwayListener
                    onClickAway={() => {
                      setOpenEmoji(false)
                    }}>
                    <div className={classes.picker}>
                      <Picker
                        /* eslint-disable */
                        onEmojiClick={(e, emoji) => {
                          setHtmlMessage(htmlMessage => htmlMessage + emoji.emoji)
                          setMessage(message + emoji.emoji)
                          setOpenEmoji(false)
                        }}
                        /* eslint-enable */
                      />
                    </div>
                  </ClickAwayListener>
                )}
              </Grid>
            </Grid>
          </ClickAwayListener>
        </Grid>
        <ChannelInputInfoMessage
          showInfoMessage={inputState !== INPUT_STATE.AVAILABLE}
          inputState={inputState}
        />
      </Grid>
    </Grid>
  )
}

export default ChannelInputComponent
