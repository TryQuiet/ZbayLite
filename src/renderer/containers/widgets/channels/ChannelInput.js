import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import ChannelInputComponent from '../../../components/widgets/channels/ChannelInput'
import channelHandlers from '../../../store/handlers/channel'
import messagesHandlers from '../../../store/handlers/messages'
import messagesQueueHandlers from '../../../store/handlers/messagesQueue'
import mentionsHandlers from '../../../store/handlers/mentions'
import channelSelectors from '../../../store/selectors/channel'
import usersSelectors from '../../../store/selectors/users'
import { MESSAGE_SIZE } from '../../../zbay/transit'

export const mapStateToProps = state => {
  return {
    message: channelSelectors.message(state),
    id: channelSelectors.id(state),
    inputState: channelSelectors.inputLocked(state),
    members: channelSelectors.members(state),
    channelName: channelSelectors.data(state)
      ? channelSelectors.data(state).username
      : ' Unnamed',
    users: usersSelectors.users(state),
    myUser: usersSelectors.myUser(state),
    isSizeCheckingInProgress: channelSelectors.isSizeCheckingInProgress(state),
    isMessageTooLong: channelSelectors.messageSizeStatus(state)
  }
}

export const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      onChange: channelHandlers.actions.setMessage,
      resetDebounce: messagesQueueHandlers.epics.resetMessageDebounce,
      sendOnEnter: channelHandlers.epics.sendOnEnter,
      checkMentions: mentionsHandlers.epics.checkMentions,
      checkMessageSizeLimit: messagesHandlers.epics.checkMessageSize
    },
    dispatch
  )
}
export const ChannelInput = ({
  onChange,
  setTab,
  sendOnEnter,
  message,
  inputState,
  channelName,
  resetDebounce,
  users,
  members,
  checkMentions,
  myUser,
  isPublicChannel,
  isMessageTooLong,
  isSizeCheckingInProgress,
  id
}) => {
  const [infoClass, setInfoClass] = React.useState(null)
  const [anchorEl, setAnchorEl] = React.useState({})
  const [mentionsToSelect, setMentionsToSelect] = React.useState([])
  return (
    <ChannelInputComponent
      infoClass={infoClass}
      setInfoClass={setInfoClass}
      id={id}
      users={users}
      onChange={e => {
        onChange({ value: e, id })
        resetDebounce()
      }}
      onKeyPress={(e) => {
        checkMentions()
        sendOnEnter(e, setTab)
      }}
      isPublicChannel={isPublicChannel}
      message={message}
      inputState={inputState}
      inputPlaceholder={`#${channelName} as @${myUser.nickname}`}
      channelName={channelName}
      messageLimit={MESSAGE_SIZE}
      anchorEl={anchorEl}
      setAnchorEl={setAnchorEl}
      mentionsToSelect={mentionsToSelect}
      setMentionsToSelect={setMentionsToSelect}
      members={members}
      isMessageTooLong={isMessageTooLong}
      isSizeCheckingInProgress={isSizeCheckingInProgress}
    />
  )
}

export default connect(mapStateToProps, mapDispatchToProps)(ChannelInput)
