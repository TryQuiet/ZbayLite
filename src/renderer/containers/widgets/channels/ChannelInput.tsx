import React, { useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import ChannelInputComponent from '../../../components/widgets/channels/ChannelInput'
import channelHandlers from '../../../store/handlers/channel'
import mentionsHandlers from '../../../store/handlers/mentions'
import channelSelectors from '../../../store/selectors/channel'
import { publicChannels, messages, identity, users } from '@zbayapp/nectar'

export const useChannelInputData = () => {
  const currentChannel = useSelector(publicChannels.selectors.currentChannel)
  const channels = useSelector(publicChannels.selectors.publicChannels)

  const data = {
    message: useSelector(channelSelectors.message),
    id: useSelector(channelSelectors.id),
    inputState: useSelector(channelSelectors.inputLocked),
    members: useSelector(channelSelectors.members),
    channelName: channels.find(channel => channel.address === currentChannel)?.name,
    users: useSelector(users.selectors.certificatesMapping),
    myUser: useSelector(identity.selectors.currentIdentity),
    isMessageTooLong: useSelector(channelSelectors.messageSizeStatus)
  }
  return data
}

export const useChannelInputActions = () => {
  const dispatch = useDispatch()

  const onChange = useCallback((arg: { value: string; id: string }) => {
    dispatch(channelHandlers.actions.setMessage(arg))
  }, [dispatch])

  const resetDebounce = useCallback(() => {
    // dispatch(messagesQueueHandlers.epics.resetMessageDebounce())
  }, [dispatch])

  const sendOnEnter = useCallback((message) => {
    console.log('sendOnEnter channel input')
    dispatch(messages.actions.sendMessage(message))
  }, [dispatch])

  const checkMentions = useCallback(() => {
    dispatch(mentionsHandlers.epics.checkMentions())
  }, [dispatch])

  return { onChange, resetDebounce, sendOnEnter, checkMentions }
}

export const ChannelInput = () => {
  const [infoClass, setInfoClass] = React.useState<string>(null)
  // eslint-disable-next-line
  const [anchorEl, setAnchorEl] = React.useState({} as HTMLElement)
  const [mentionsToSelect, setMentionsToSelect] = React.useState<any[]>([])

  const { channelName, id, inputState, isMessageTooLong, members, message, myUser, users } = useChannelInputData()
  const { checkMentions, onChange, resetDebounce, sendOnEnter } = useChannelInputActions()

  return (
    <ChannelInputComponent
      infoClass={infoClass}
      setInfoClass={setInfoClass}
      id={id}
      users={users}
      onChange={e => {
        onChange({ value: e, id })
        resetDebounce()
      }}
      onKeyPress={(message) => {
        console.log(message, 'message is')
        checkMentions()
        sendOnEnter(message)
      }}
      message={message}
      inputState={inputState}
      inputPlaceholder={`#${channelName} as @${myUser ? myUser.zbayNickname : ''}`}
      channelName={channelName}
      anchorEl={anchorEl}
      setAnchorEl={setAnchorEl}
      mentionsToSelect={mentionsToSelect}
      setMentionsToSelect={setMentionsToSelect}
      members={members}
      isMessageTooLong={isMessageTooLong}
    />
  )
}

export default ChannelInput
