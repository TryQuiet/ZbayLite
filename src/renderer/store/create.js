import { applyMiddleware, createStore, compose } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const composer = (enhancers) => {
  if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ &&
      process.env.NODE_ENV === 'development') {
    return window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__(enhancers)
  }
  return compose(enhancers)
}

export default ({
  initialState
} = {}) => createStore(
  reducers,
  initialState,
  composer(applyMiddleware(...[errorsMiddleware, createDebounce(), thunk, promise()]))
)
