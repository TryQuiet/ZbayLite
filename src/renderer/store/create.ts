import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'
import promise from 'redux-promise-middleware'
import createDebounce from 'redux-debounced'
import createSagaMiddleware from 'redux-saga'
import rootSaga from '../sagas/index.saga'
import { composeWithDevTools } from 'redux-devtools-extension'

import reducers from './reducers'
import { errorsMiddleware } from './middlewares'

const sagaMiddleware = createSagaMiddleware()

export default (initialState = {}) => {
  const store = createStore(
    reducers,
    initialState,
    composeWithDevTools(
      applyMiddleware(...[errorsMiddleware, createDebounce(), sagaMiddleware, thunk, promise()])
    )
  )
  sagaMiddleware.run(rootSaga)
  return store
}
