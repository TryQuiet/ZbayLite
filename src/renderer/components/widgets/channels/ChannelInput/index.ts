import ChannelInput from './ChannelInput'

export default ChannelInput
