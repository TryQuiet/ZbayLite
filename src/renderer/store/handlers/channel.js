import Immutable from 'immutable'
import BigNumber from 'bignumber.js'
import { createAction, handleActions } from 'redux-actions'
import crypto from 'crypto'
import * as R from 'ramda'
import { remote } from 'electron'
import { DateTime } from 'luxon'

import history from '../../../shared/history'
import operationsHandlers from './operations'
import notificationsHandlers from './notifications'
// import messagesQueueHandlers from './messagesQueue'
import messagesHandlers, { _checkMessageSize } from './messages'
import channelsHandlers from './channels'
import offersHandlers from './offers'
import channelSelectors from '../selectors/channel'
import identitySelectors from '../selectors/identity'
import contactsSelectors from '../selectors/contacts'
import client from '../../zcash'
import { messages } from '../../zbay'
import { errorNotification, LoaderState } from './utils'
import { messageType, actionTypes } from '../../../shared/static'
import { DisplayableMessage } from '../../zbay/messages'
import usersSelectors from '../selectors/users'
import contactsHandlers from './contacts'
import electronStore from '../../../shared/electronStore'
import { channelToUri } from '../../../renderer/zbay/channels'

export const ChannelState = Immutable.Record(
  {
    spentFilterValue: new BigNumber(0),
    id: null,
    message: Immutable.Map(),
    shareableUri: '',
    address: '',
    loader: LoaderState({ loading: false }),
    members: null,
    showInfoMsg: true,
    isSizeCheckingInProgress: false,
    messageSizeStatus: null,
    displayableMessageLimit: 50
  },
  'ChannelState'
)

export const initialState = ChannelState()

const setSpentFilterValue = createAction(
  actionTypes.SET_SPENT_FILTER_VALUE,
  (_, value) => value
)
const setMessage = createAction(actionTypes.SET_CHANNEL_MESSAGE)
const setDisplayableLimit = createAction(actionTypes.SET_DISPLAYABLE_LIMIT)
const setChannelId = createAction(actionTypes.SET_CHANNEL_ID)
const setLoading = createAction(actionTypes.SET_CHANNEL_LOADING)
const setLoadingMessage = createAction(actionTypes.SET_CHANNEL_LOADING_MESSAGE)
const setShareableUri = createAction(actionTypes.SET_CHANNEL_SHAREABLE_URI)
const setAddress = createAction(actionTypes.SET_CHANNEL_ADDRESS)
const resetChannel = createAction(actionTypes.SET_CHANNEL)
const isSizeCheckingInProgress = createAction(
  actionTypes.IS_SIZE_CHECKING_IN_PROGRESS
)
const messageSizeStatus = createAction(actionTypes.MESSAGE_SIZE_STATUS)

export const actions = {
  setLoading,
  setLoadingMessage,
  setSpentFilterValue,
  setMessage,
  setShareableUri,
  setChannelId,
  resetChannel,
  isSizeCheckingInProgress,
  messageSizeStatus,
  setDisplayableLimit
}

const loadChannel = key => async (dispatch, getState) => {
  try {
    dispatch(setChannelId(key))
    dispatch(setDisplayableLimit(30))
    // Calculate URI on load, that way it won't be outdated, even if someone decides
    // to update channel in vault manually
    const contact = contactsSelectors.contact(key)(getState())
    const unread = contact.newMessages.size
    remote.app.badgeCount = remote.app.badgeCount - unread
    const ivk = electronStore.get(
      `defaultChannels.${contact.get('address')}.keys.ivk`
    )
    if (ivk) {
      const uri = await channelToUri({
        name: contact.get('username'),
        ivk
      })
      dispatch(setShareableUri(uri))
    }
    electronStore.set(
      `lastSeen.${key}`,
      parseInt(DateTime.utc().toSeconds()).toString()
    )
    dispatch(setAddress(contact.address))

    dispatch(contactsHandlers.actions.cleanNewMessages({ contactAddress: key }))
    // await dispatch(clearNewMessages())
    // await dispatch(updateLastSeen())
  } catch (err) {}
}
const loadOffer = (id, address) => async (dispatch, getState) => {
  try {
    await dispatch(offersHandlers.epics.updateLastSeen({ itemId: id }))
    dispatch(setChannelId(id))
    dispatch(setShareableUri(''))
    dispatch(setAddress(address))
  } catch (err) {}
}
const linkChannelRedirect = targetChannel => async (dispatch, getState) => {
  const contact = contactsSelectors.contact(targetChannel.address)(getState())
  if (contact.address) {
    history.push(`/main/channel/${targetChannel.address}`)
    return
  }

  // We can parse timestamp to blocktime and get accurate birthday block for this channel
  // Skipped since we dont support rescaning also we already got birthday of zbay as main wallet birthday
  await client.importKey(targetChannel.keys.ivk)
  await dispatch(
    contactsHandlers.actions.addContact({
      key: targetChannel.address,
      contactAddress: targetChannel.address,
      username: targetChannel.name
    })
  )
  const importedChannels = electronStore.get(`importedChannels`) || {}
  electronStore.set('importedChannels', {
    ...importedChannels,
    [targetChannel.address]: {
      address: targetChannel.address,
      name: targetChannel.name,
      description: targetChannel.description,
      owner: targetChannel.owner,
      keys: targetChannel.keys
    }
  })
  history.push(`/main/channel/${targetChannel.address}`)
}
const sendOnEnter = (event, resetTab) => async (dispatch, getState) => {
  if (resetTab) {
    resetTab(0)
  }
  const enterPressed = event.nativeEvent.keyCode === 13
  const shiftPressed = event.nativeEvent.shiftKey === true
  const channel = channelSelectors.channel(getState()).toJS()
  const messageToSend = channelSelectors.message(getState())
  let message
  if (enterPressed && !shiftPressed) {
    event.preventDefault()
    const privKey = identitySelectors.signerPrivKey(getState())
    message = messages.createMessage({
      messageData: {
        type: messageType.BASIC,
        data: messageToSend
      },
      privKey: privKey
    })
    const isMergedMessageTooLong = await dispatch(
      _checkMessageSize(message.message)
    )
    if (!isMergedMessageTooLong) {
      dispatch(setMessage(''))
      const myUser = usersSelectors.myUser(getState())
      const messageDigest = crypto.createHash('sha256')

      const messageEssentials = R.pick(['createdAt', 'message', 'spent'])(
        message
      )
      const key = messageDigest
        .update(JSON.stringify(messageEssentials))
        .digest('hex')

      const messagePlaceholder = DisplayableMessage({
        ...message,
        id: key,
        sender: {
          replyTo: myUser.address,
          username: myUser.nickname
        },
        fromYou: true,
        status: 'pending',
        message: messageToSend
      })
      dispatch(
        contactsHandlers.actions.addMessage({
          key: channel.id,
          message: { [key]: messagePlaceholder }
        })
      )
      dispatch(
        operationsHandlers.actions.addOperation({
          channelId: channel.id,
          id: key
        })
      )
      const identityAddress = identitySelectors.address(getState())

      const transfer = await messages.messageToTransfer({
        message: message,
        address: channel.address,
        identityAddress
      })
      const transaction = await client.sendTransaction(transfer)
      console.log(transaction, 'transaction details')
      if (!transaction.txid) {
        dispatch(
          contactsHandlers.actions.addMessage({
            key: channel.id,
            message: { [key]: messagePlaceholder.set('status', 'failed') }
          })
        )
        dispatch(
          notificationsHandlers.actions.enqueueSnackbar(
            errorNotification({
              message:
                "Couldn't send the message, please check node connection."
            })
          )
        )
        return
      }
      dispatch(
        operationsHandlers.epics.resolvePendingOperation({
          channelId: channel.id,
          id: key,
          txid: transaction.txid
        })
      )
    }
  }
}
const sendChannelSettingsMessage = ({
  address,
  minFee = '0',
  onlyRegistered = '0'
}) => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const owner = identitySelectors.signerPubKey(getState())
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageType.CHANNEL_SETTINGS,
      data: {
        owner,
        minFee,
        onlyRegistered
      }
    },
    privKey: privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: address,
    identityAddress
  })
  try {
    const txid = await client.sendTransaction(transfer)
    if (txid.error) {
      throw new Error(txid.error)
    }
    return 1
  } catch (err) {
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't create channel, please check node connection."
        })
      )
    )
    return -1
  }
}

const resendMessage = messageData => async (dispatch, getState) => {
  const identityAddress = identitySelectors.address(getState())
  const channel = channelSelectors.data(getState()).toJS()
  const privKey = identitySelectors.signerPrivKey(getState())
  const message = messages.createMessage({
    messageData: {
      type: messageData.type,
      data: messageData.message,
      spent: parseFloat(messageData.spent.toString())
    },
    privKey
  })
  const transfer = await messages.messageToTransfer({
    message,
    address: channel.address,
    amount: parseFloat(messageData.spent.toString()),
    identityAddress
  })
  const messagePlaceholder = DisplayableMessage({
    ...messageData,
    status: 'pending'
  })
  dispatch(
    contactsHandlers.actions.addMessage({
      key: channel.key,
      message: { [messageData.id]: messagePlaceholder }
    })
  )
  const transaction = await client.sendTransaction(transfer)
  if (!transaction.txid) {
    dispatch(
      contactsHandlers.actions.addMessage({
        key: channel.key,
        message: {
          [messageData.id]: messagePlaceholder.set('status', 'failed')
        }
      })
    )
    dispatch(
      notificationsHandlers.actions.enqueueSnackbar(
        errorNotification({
          message: "Couldn't send the message, please check node connection."
        })
      )
    )
    return
  }
  dispatch(
    operationsHandlers.epics.resolvePendingOperation({
      channelId: channel.key,
      id: messageData.id,
      txid: transaction.txid
    })
  )
}

const updateLastSeen = () => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  return dispatch(channelsHandlers.epics.updateLastSeen({ channelId }))
}

const clearNewMessages = () => async (dispatch, getState) => {
  const channelId = channelSelectors.channelId(getState())
  dispatch(messagesHandlers.actions.cleanNewMessages({ channelId }))
}

export const epics = {
  sendOnEnter,
  loadChannel,
  resendMessage,
  clearNewMessages,
  updateLastSeen,
  loadOffer,
  sendChannelSettingsMessage,
  linkChannelRedirect
}

// TODO: we should have a global loader map
export const reducer = handleActions(
  {
    [setLoading]: (state, { payload: loading }) =>
      state.setIn(['loader', 'loading'], loading),
    [setLoadingMessage]: (state, { payload: message }) =>
      state.setIn(['loader', 'message'], message),
    [setSpentFilterValue]: (state, { payload: value }) =>
      state.set('spentFilterValue', new BigNumber(value)),
    [setMessage]: (state, { payload: value }) =>
      state.setIn(['message', state.get('id')], value),
    [setChannelId]: (state, { payload: id }) => state.set('id', id),
    [isSizeCheckingInProgress]: (state, { payload }) =>
      state.set('isSizeCheckingInProgress', payload),
    [messageSizeStatus]: (state, { payload }) =>
      state.set('messageSizeStatus', payload),
    [setShareableUri]: (state, { payload: uri }) =>
      state.set('shareableUri', uri),
    [setDisplayableLimit]: (state, { payload: limit }) =>
      state.set('displayableMessageLimit', limit),
    [setAddress]: (state, { payload: address }) =>
      state.set('address', address),
    [resetChannel]: () => initialState
  },
  initialState
)

export default {
  reducer,
  epics,
  actions
}
