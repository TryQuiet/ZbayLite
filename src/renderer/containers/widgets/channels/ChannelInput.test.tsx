import BigNumber from 'bignumber.js'
import { renderHook } from '@testing-library/react-hooks'
import React from 'react'
import { useChannelInputData, useChannelInputActions } from './ChannelInput'

import { Provider } from 'react-redux'

import create from '../../../store/create'
import { ChannelState } from '../../../store/handlers/channel'
import { createChannel } from '../../../testUtils'

const channelId = 1234
describe('ChannelInput', () => {
  let store, wrapper

  beforeEach(() => {
    jest.clearAllMocks()
    store = create({
      channel: {
        ...ChannelState,
        spentFilterValue: 38,
        name: 'Politics',
        members: new BigNumber(0),
        message: 'This is a test message',
        messages: [],
        id: channelId
      },
      channels: {
        data: [createChannel(channelId)]
      }
    })
    wrapper = ({ children }) => <Provider store={store}>{children}</Provider>
  })

  it('will receive right props', async () => {
    const { result } = renderHook(() => useChannelInputData(), { wrapper })
    expect(result.current).toMatchSnapshot()
  })

  it('will receive right actions', async () => {
    const { result } = renderHook(() => useChannelInputActions(), { wrapper })
    expect(result.current).toMatchSnapshot()
  })
})
