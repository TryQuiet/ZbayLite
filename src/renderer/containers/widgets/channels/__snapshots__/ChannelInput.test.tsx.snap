// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelInput will receive right actions 1`] = `
Object {
  "checkMentions": [Function],
  "onChange": [Function],
  "resetDebounce": [Function],
  "sendOnEnter": [Function],
}
`;

exports[`ChannelInput will receive right props 1`] = `
Object {
  "channelName": undefined,
  "id": 1234,
  "inputState": 0,
  "isMessageTooLong": null,
  "members": Set {},
  "message": "",
  "myUser": undefined,
  "users": Object {},
}
`;
